dev-tools/xtask/src/external.rs

```diff
@@ -0,0 +1,68 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! External xtasks. (extasks?)
+
+use std::ffi::{OsStr, OsString};
+use std::os::unix::process::CommandExt;
+use std::process::Command;
+
+use anyhow::{Context, Result};
+use clap::Parser;
+
+/// Argument parser for external xtasks.
+///
+/// In general we want all developer tasks to be discoverable simply by running
+/// `cargo xtask`, but some development tools end up with a particularly
+/// large dependency tree. It's not ideal to have to pay the cost of building
+/// our release engineering tooling if all the user wants to do is check for
+/// workspace dependency issues.
+///
+/// `External` provides a pattern for creating xtasks that live in other crates.
+/// An external xtask is defined on `crate::Cmds` as a tuple variant containing
+/// `External`, which captures all arguments and options (even `--help`) as
+/// a `Vec<OsString>`. The main function then calls `External::exec` with the
+/// appropriate bin target name and any additional Cargo arguments.
+#[derive(Parser)]
+#[clap(
+    disable_help_flag(true),
+    disable_help_subcommand(true),
+    disable_version_flag(true)
+)]
+pub struct External {
+    #[clap(trailing_var_arg(true), allow_hyphen_values(true))]
+    args: Vec<OsString>,
+
+    #[clap(skip)]
+    command: Option<Command>,
+}
+
+impl External {
+    pub fn cargo_args(
+        mut self,
+        args: impl IntoIterator<Item = impl AsRef<OsStr>>,
+    ) -> External {
+        self.command.get_or_insert_with(default_command).args(args);
+        self
+    }
+
```

Comment:
Rather than calling `cargo_args` and `exec`, we could just call:

```rust
pub struct External {
  args: Vec<OsString>,
}

impl External {
  pub fn exec(self, binary: impl AsRef<OsStr>) -> Result<()> {
    default_command()  // I might rename this to "cargo_run_command()"
      .arg("--bin")
      .arg(binary)
      .arg("--")
      .args(self.args)
      .exec();
    ...
  }
}
```

I think this has the same functionality, but makes it a little clear that the "default" command is not actually modifiable.

Reply:
Yeah, this is what I was wondering too -- any particular reason to have `command` be part of the struct rather than creating it here?

Reply:
What this is missing is the ability to add `--release` to the command, which is particularly useful for the releng tooling. (I don't expect that all external xtasks should be built and run in release mode.)

Reply:
Specifically I imagine the usual flow in main.rs would be:

```rust
Cmds::Whatever(external) => external.exec("omicron-whatever"),
```

What the `Option<Command>`, `defualt_command`, etc chicanery gets us is the ability to optionally add additional arguments to cargo itself, which we use for `Cmds::Releng`.